### ISSUE TITLE
[BUG]: Add test case for nvbug 4678853

### ISSUE THREAD

**Issue by wmaxey** [NVIDIA/cccl#2045]
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Compile-time Error

### Component

CUB

### Describe the bug

Multiple definition error from 4678853 needs a reproducer added to our test suite. This should be as simple as linking together a pair of each header. We should probably try to do this in every project if not already done.

### How to Reproduce

```
create bar.cc
#include <thrust/complex.h>

create foo.cc
#include <thrust/complex.h>
```

```
build:
g++ -c -fPIC -I /usr/local/cuda/include bar.cc
g++ -c -fPIC -I /usr/local/cuda/include foo.cc
g++ -fPIC bar.o foo.o -shared -Wl,-soname,libhello.so -o libhello.so
```

### Expected behavior

We shouldn't emit multiple definition errors by leaking non-static interfaces into the TU.

**Comment by miscco** [comment]
I believe we should do that in the context of our all public headers test set.

We already have all public headers there and it should be simple to create two TUs that include all those headers and link two.

**Comment by wmaxey** [comment]
> We already have all public headers there and it should be simple to create two TUs that include all those headers and link two.

Exactly. We just need to flex that we can link all the permutations of NVCC + compiler. 

e.g.
NVCC+NVCC
GCC+NVCC
GCC+GCC.
